dns.tf

```diff
@@ -7,3 +7,58 @@ data "azurerm_resource_group" "proddns_jenkinsio" {
 data "azurerm_dns_zone" "jenkinsio" {
   name = "jenkins.io"
 }
+
+resource "azurerm_dns_zone" "child_zones" {
+  for_each = local.lets_encrypt_dns_challenged_domains
+
+  name = each.key
+  # Use the same resource group for all DNS zones
+  resource_group_name = data.azurerm_resource_group.proddns_jenkinsio.name
+
+  tags = local.default_tags
+}
+
+# create DNS record of type NS for child-zone in the parent zone (to allow propagation of DNS records)
+resource "azurerm_dns_ns_record" "child_zone_ns_records" {
+  for_each = local.lets_encrypt_dns_challenged_domains
+
+  name                = trimsuffix(each.key, ".jenkins.io") # only the flat name not the fqdn
+  zone_name           = data.azurerm_dns_zone.jenkinsio.name
+  resource_group_name = data.azurerm_resource_group.proddns_jenkinsio.name
+  ttl                 = 60
+
+  records = azurerm_dns_zone.child_zones[each.key].name_servers
+}
+
+resource "azuread_application" "letsencrypt_dns_challenges" {
+  for_each = { for key, value in local.lets_encrypt_dns_challenged_domains : key => value if value == "service_principal" }
+
+  display_name = replace(each.key, ".", "_")
```

Comment:
what you refer as "terraform resource" is only a terraform-internal reference: there is nothing on the remote API (Azure in this case) named with underscores "ala terraform".

So none of the 2 choices has any impact

Reply:
Post-deployment note: we used the proposed suggestion of naming convention (using carrets `-` instead of underscores `_`) in https://github.com/jenkins-infra/azure-net/commit/77267a9de6afb28ea54085db1823ed1e6d6a0aba 

We faced an error in the deployment: as per https://learn.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#check-azure-subscription-permissions, creating an Azure AD application is not authorized with only the "Contributor" Role. As we do not want to grant "Owner" to the terraform-production Service Principal used by terraform, we moved to a data source referencing the application externall created (by an admin)

Reply:
it's nothing to do with contributor or owner, Azure AD permissions are completely separate to Azure RM.

You need to grant it either the Application Developer (can create new apps but not modify existing) / Application Administrator (full access) role or give it specific API permissions on the Graph API if you want to allow it to create via terraform.

Reply:
Thanks @timja ! We arrived to the same conlusion, and I'm currently trying https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/guides/service_principal_configuration#azure-active-directory-permissions but with no success. i'll try the "Application Developer" option.

Reply:
wip: https://github.com/jenkins-infra/azure-net/pull/32